### PR TITLE
Option for create nfs exported directory

### PR DESCRIPTION
**Pull request by bschonec**
I have a case where an application I'm installing expects that the folders to be exported via NFS do NOT exist at application install time.

This PR allows for the edge case where you don't want to create the directories but you do want to manage the NFS export file(s).

### CODE REVIEW

**Comment by tuxmea** [comment]
@bschonec Can you please rebase your branch?

**Comment by bschonec** [comment]
@tux, sure thing.  I'm not sure where I should put the tests for the directory (not) creation.  There are a lot of places in the tests where it does "ensure directory."  I only need to test for "not ensure directory."  Should I just create a simple nfsv4/nfsv3 section and do a "not ensure directory" and not test for anything else?

File: manifests/functions/create_export.pp [hunk 1]
```diff
@@ -38,13 +43,16 @@
...
+    #     # defined previously AND the $create_dir boolean is true.
     unless defined(File[$name]) {
-      file { $name:
-        ensure                  => directory,
-        owner                   => $owner,
-        group                   => $group,
-        mode                    => $mode,
-        selinux_ignore_defaults => true,
+      if $create_dir {
+        filepath { $name:
```
Comment (h-haaks) [inline manifests/functions/create_export.pp:50]: Why the switch of `file` into `filepath`? That's not relevant when it comes to creating the folder or not and there is another PR for this change.

I would argue that this module almost never should create this folder as that normally is the responsibility of another module. I think it would be even worse to let it create an entire path ...

It would be a breaking change, but I would consider setting $create_dir default to false.

File: manifests/server/export.pp [hunk 2]
```diff
@@ -57,7 +57,9 @@
 # @param v4_export_name
 # @param nfsv4_bindmount_enable
 #
-# @examples
```
Comment (tuxmea) [inline manifests/server/export.pp:60]: please re-add the `@examples` strings tag.

Comment (h-haaks) [reply manifests/server/export.pp:60]: I think it should be `@example`, and it's is a tag used by puppet strings to render an example in REFERANCE.md
See https://www.puppet.com/docs/puppet/8/puppet_strings.html

It should be moved below the last `@param` at line 65

Comment (bschonec) [reply manifests/server/export.pp:60]: Should I create a separate pull request of off the current master to fix @examples/@example?  

If so, #193

Comment (tuxmea) [reply manifests/server/export.pp:60]: No need to have another PR. please fix it here and please keep in mind that the code example must be indeted by 2 spaces.

Comment (bschonec) [reply manifests/server/export.pp:60]: I've fixed it here but I have no idea what you're referring to when you say, "...code example must be indented by 2 spaces."

Comment (kenyon) [reply manifests/server/export.pp:60]: @bschonec the example should have a description, then the example code. The code should be indented two spaces, so there should be three spaces after the hash mark `#`. Examples: https://github.com/search?q=repo%3Avoxpupuli%2Fpuppet-systemd%20%40example&type=code

Comment (tuxmea) [reply manifests/server/export.pp:60]: Tag was readded. This is OK for me. We can solve this conversation.

File: manifests/functions/create_export.pp (hunk 1 above)
Comment (tuxmea) [reply manifests/functions/create_export.pp:50]: Please stay with file for now. Besides this: there are two implementations for creating recursive directories: filepath and dirtrtee. We should have this change in a separate PR so we can check and review which module is the better one to use.

Comment (h-haaks) [reply manifests/functions/create_export.pp:50]: And the module it self has this https://github.com/voxpupuli/puppet-nfs/blob/master/manifests/functions/mkdir.pp :)

**Comment by bschonec** [comment]
I'm having trouble figuring out where to put the "if create_dir = true" tests.  Should the test be in defines/server_export_spec.rb or in classes/nfs_spec.rb?

File: manifests/functions/create_export.pp [hunk 3]
```diff
@@ -34,13 +39,16 @@
...
-        group                   => $group,
-        mode                    => $mode,
-        selinux_ignore_defaults => true,
+      if $create_dir {
+        file { $name:
+          ensure => directory,
+          owner  => $owner,
+          group  => $group,
+          mode   => $mode,
+        }
```
Comment (smortex) [inline manifests/functions/create_export.pp:51]: Given it will not manage the resource, I suggest to name the new param `manage` rather than `create` (i.e. the file resource create a non-existing directory but also ensure owner, group, permissions, etc).

Comment (bschonec) [reply manifests/functions/create_export.pp:51]: If $create_dir is true (which is the default) then the directory is managed.  Or am misunderstanding your comment?  My use case here is that I have a 3rd party application that expects the directory to be absent.  The application install will fail if the directory is created but I still have to export the directory via NFS.  This solves the problem of exporting the directory but not being required to create the directory/export.

File: manifests/functions/create_export.pp [hunk 4]
```diff
@@ -34,13 +39,16 @@
...
 
+    # Create the directory path only if a File resource isn't
+    #     # defined previously AND the $create_dir boolean is true.
     unless defined(File[$name]) {
-      file { $name:
-        ensure                  => directory,
-        owner                   => $owner,
-        group                   => $group,
-        mode                    => $mode,
-        selinux_ignore_defaults => true,
```
Comment (bastelfreak) [inline manifests/functions/create_export.pp:43]: Why did you delete this option?

Comment (bschonec) [reply manifests/functions/create_export.pp:43]: A fubar in my brain.  It's back now.

File: manifests/functions/create_export.pp [hunk 5]
```diff
@@ -15,13 +15,18 @@
 # @param mode
 #   Sets the permissions of the exported directory.
 #
+# @param create_dir
+#   Boolean.  Create the directory to be exported.
```
Comment (wyardley) [inline manifests/functions/create_export.pp:19]: nit, but the type and default get added when you build `Reference.md` using puppet-strings, so both should be omitted here. You might want to reword it to something along the lines of:
```suggestion
#   Whether or not to create the directory to be exported.
```
(and then remove line 20 entirely).

File: manifests/server/export.pp [hunk 6]
```diff
@@ -47,6 +47,10 @@
 # @param v4_export_name
 # @param nfsv4_bindmount_enable
 #
+# @param create_dir
+#   Boolean.  Create the directory to be exported.
+#   Defaults to true.
```
Comment (wyardley) [inline manifests/server/export.pp:52]: same feedback as the other parameter with generated docs.

File: manifests/functions/create_export.pp (hunk 3 above)
Comment (smortex) [reply manifests/functions/create_export.pp]: Sorry to be unclear, I am suggesting naming the parameter `manage_dir` rather than `create_dir`.

Or `manage_directory` as bytes are cheap nowadays :smile: 

Because if you set `manage_directory => false` and you put the wrong owner / group on the directory you created by hand, puppet will not fix this because it does not manage that directory.

File: manifests/server/export.pp (hunk 6 above)
Comment (bschonec) [reply manifests/server/export.pp]: Changed per your recommendations.  Thanks.

File: manifests/functions/create_export.pp (hunk 3 above)
Comment (bschonec) [reply manifests/functions/create_export.pp]: Changed per your recommendations.  Thanks.

File: manifests/server/export.pp [hunk 7]
```diff
@@ -55,6 +58,15 @@
...
 #   }
 #
+# nfs::server::export { '/srv/nfs_exported/directory':
+#   clients => '1.2.3.4/24(rw,insecure,no_subtree_check,async,no_root_squash) 5.6.7.8/24(ro)',
+#   share => 'share_name_on_nfs_server',
+# }
+#
+# Links
+#
+# * {Puppet Docs: Using Parameterized Classes}[http://j.mp/nVpyWY]
```
Comment (bastelfreak) [inline manifests/server/export.pp]: can you use the original link? the shortner is dead. puppet-strings also has a real annotation:

`# @see my awesome description https://blablabla.de`

Comment (bastelfreak) [reply manifests/server/export.pp]: edit: maybe first the link, then the description :thinking:
